### ISSUE TITLE
[Flaky test] "PersistentVolumeClaims not all in phase Bound within 5m0s" failing in ~2% of kind presubmits

### ISSUE THREAD

**Issue by liggitt** [kubernetes/kubernetes#93358]
**Which jobs are flaking**:

* pull-kubernetes-e2e-kind-ipv6
* pull-kubernetes-e2e-kind

**Which test(s) are flaking**:

* CSI mock volume ... tests

**Testgrid link**:

* https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-e2e-kind&include-filter-by-regex=CSI%20mock%20volume&width=5&sort-by-failures=
* https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-e2e-kind-ipv6&include-filter-by-regex=CSI%20mock%20volume&width=5&sort-by-failures=

**Reason for failure**:

```
test/e2e/storage/csi_mock_volume.go:351
Jul 22 00:33:25.985: Failed waiting for PVC to be bound PersistentVolumeClaims [pvc-7fnn6] not all in phase Bound within 5m0s
Unexpected error:
    <*errors.errorString | 0xc002072890>: {
        s: "PersistentVolumeClaims [pvc-7fnn6] not all in phase Bound within 5m0s",
    }
    PersistentVolumeClaims [pvc-7fnn6] not all in phase Bound within 5m0s
occurred
test/e2e/storage/csi_mock_volume.go:1200
```

**Anything else we need to know**:

* https://storage.googleapis.com/k8s-gubernator/triage/index.html?ci=0&pr=1&text=not%20all%20in%20phase%20Bound%20within%205m0s&test=CSI%20mock%20volume

Example failures:
* https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-kind/1286019551993008128
* https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-kind/1285857244977565699
* https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-kind/1285673283571683329

/sig storage

**Comment by liggitt** [comment]
might be worth looking at the average bind time in successful run cases over time and seeing how close to the 5 minute timeout we are... what is the expected time to bind in these tests?

**Comment by BenTheElder** [comment]
I'd love to know what is happening during bind, exactly. Can anyone give a quick overview? I can dig in later but it would help me determine if there's something we can do to mitigate in kind quicker.

**Comment by liggitt** [comment]
cc @kubernetes/sig-storage-test-failures 

/assign @msau42 
for routing

**Comment by msau42** [comment]
normally creating a pvc and waiting for it to bind calls out to the storage provider to create a disk, which could take time. However, this is csi mock driver, where the create volume returns almost immediately. So that implies the slowdown would be in the controllers.

/assign @Jiawei0227
to help dig in further where the time is being spent

**Comment by msau42** [comment]
cc @jiawei0227

**Comment by Jiawei0227** [comment]
Some findings:

1. The reason why we got a 5 minute timeout issue is that the csi-provisioner starts after the timeout. To be exact:
In the case of [this failure](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-kind/1286019551993008128):
```Jul 22 19:49:33.768: INFO: Waiting up to 5m0s for PersistentVolumeClaims [pvc-wz95b] to have phase Bound```
The testing system starts to expect the pvc to be created at 19:49. However, by looking at the `csi-provisioner` logs we can find:

```
I0722 19:51:03.234109       1 feature_gate.go:243] feature gates: &{map[Topology:true]}
I0722 19:51:03.234149       1 csi-provisioner.go:107] Version: v2.0.0-rc1-0-g94aaacb09
I0722 19:51:03.234171       1 csi-provisioner.go:121] Building kube configs for running in cluster...
I0722 19:51:03.265650       1 connection.go:153] Connecting to unix:///csi/csi.sock
W0722 19:51:13.272583       1 connection.go:172] Still connecting to unix:///csi/csi.sock
W0722 19:51:23.276212       1 connection.go:172] Still connecting to unix:///csi/csi.sock
W0722 19:51:33.299606       1 connection.go:172] Still connecting to unix:///csi/csi.sock
...(similar)
W0722 19:54:23.271699       1 connection.go:172] Still connecting to unix:///csi/csi.sock
W0722 19:54:33.268121       1 connection.go:172] Still connecting to unix:///csi/csi.sock
I0722 19:54:38.999170       1 common.go:111] Probing CSI driver for readiness
I0722 19:54:38.999214       1 connection.go:182] GRPC call: /csi.v1.Identity/Probe
I0722 19:54:38.999222       1 connection.go:183] GRPC request: {}
I0722 19:54:39.002492       1 connection.go:185] GRPC response: {"ready":{"value":true}}
```

The CSI-provisioner only gets to connect to the socket on 19:54, which is after the 5 minute timeout. That's the reason the test case failed. However, it is weird that it took so long for the `mock-csi-driver` container to start the socket. I took a look at `kubelet` and `containerd` log and couldn't find something obviously wrong. It seems we don't keep much information about the pod id and container id of the csi driver so it is hard to identify the corresponding log. I quickly looked some other passed tests, it is normal to take about 3~ minutes for the `csi-provisioner` talk to the socket. So that is one thing we might need to further dig out. Something we can do first is to add timestamps to mock driver log so that we have a better idea of the timeline.

2. The third flaky test [failure example](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-kind/1285673283571683329) might be a similar issue with #91001
I have found the following log in mock csi driver:

```gRPCCall: {"Method":"/csi.v1.Node/NodeUnpublishVolume","Request":{"volume_id":"4","target_path":"/var/lib/kubelet/pods/9b15a657-fbbe-439e-8e44-4606aed57d88/volumes/kubernetes.io~csi/pvc-677af66b-b0c9-4e3a-bf66-8e215a57f4ad/mount"},"Response":{},"Error":"","FullError":null}
gRPCCall: {"Method":"/csi.v1.Controller/DeleteVolume","Request":{"volume_id":"4"},"Response":{},"Error":"","FullError":null}
gRPCCall: {"Method":"/csi.v1.Node/NodeGetCapabilities","Request":{},"Response":{"capabilities":[{"Type":{"Rpc":{}}},{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}}]},"Error":"","FullError":null}
gRPCCall: {"Method":"/csi.v1.Node/NodeUnstageVolume","Request":{"volume_id":"4","staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-677af66b-b0c9-4e3a-bf66-8e215a57f4ad/globalmount"},"Response":null,"Error":"rpc error: code = NotFound desc = 4","FullError":{"code":5,"message":"4"}}
gRPCCall: {"Method":"/csi.v1.Node/NodeGetCapabilities","Request":{},"Response":{"capabilities":[{"Type":{"Rpc":{}}},{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}}]},"Error":"","FullError":null}}
```

It seems like DeleteVolume happens before NodeUnstageVolume? I am not  sure if this is the expected behavior.  But if it is not, it might lead to some error status.

**Comment by msau42** [comment]
So it seems like starting the Pod is slow for some reason.  @BenTheElder do you know if we can increase the kubelet log level in kind clusters so we can get more details about how long it's taking to bring up the pods?

**Comment by BenTheElder** [comment]
@aojea recently made the log level adjustable in CI, https://github.com/kubernetes-sigs/kind/pull/1750

**Comment by BenTheElder** [comment]
Are we starting a new mock driver every time with multi-minute startup?

**Comment by msau42** [comment]
Yes every test case brings up the a new mock driver pod.  However, I'm not clear on where the time is being spent, is it kubelet taking a long time to start the container, or is there something in the mock driver startup sequence that is taking a long time. Unfortunately we couldn't find any kubelet logs about the container, and the mock driver logs themselves don't have timestamps.

The mock driver pod does include the csi-provisioner container, and those logs do have timestamps. And we see that the csi-provisioner container started initializing after the test already timed out, so it does indicate to me that kubelet is slow to bring up the pod in general.

**Comment by BenTheElder** [comment]
We can increase the log verbosity (see above), I don't know what value to increase them to though.

How fast does this pod come up elsewhere? Do we need to create a new one every time?

I don't think pod startup normally takes O(5 min) on kind ... is there some kind of readiness gate or something?

**Comment by msau42** [comment]
v=4 is the usual level I see for test clusters.

There are no readiness gates: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/testing-manifests/storage-csi/mock/csi-mock-driver.yaml. To me it seems like the slowdown is in the bringup of the container itself, not the actual application.

For the mock driver tests, we toggle a lot of different behaviors/features on the driver per test case, so we bring up a unique mock driver per test case with different configs.

@Jiawei0227 can you look into the timings of mock driver bringup on other ci jobs, like https://k8s-testgrid.appspot.com/sig-release-1.19-blocking#gce-cos-k8sbeta-default&include-filter-by-regex=CSI.*mock

**Comment by BenTheElder** [comment]
kind runs should be v=4 since 4 days ago https://github.com/kubernetes-sigs/kind/pull/1750/files#r461102826

**Comment by Jiawei0227** [comment]
I took a quick round look at the mock driver bringup time on these ci(about 10). https://k8s-testgrid.appspot.com/sig-release-1.19-blocking#gce-cos-k8sbeta-default&include-filter-by-regex=CSI.*mock. What I find is that: first of all, this is e2e running on gce, which in my guess is a lot different from kind. Second, the bringup time of mock driver csi on GCE is a lot faster than kind. The average time I looked is about 5s or so. That's a pretty reasonable time for csi-provisioner to talk to the socket.

I am suspecting if there is anything weird in kind cluster? I am going to take another look at e2e on kind cluster and report back the average time.

**Comment by Jiawei0227** [comment]
Another interesting case I found: [failure](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-kind/1289150375802179587)
1. The mock CSI driver never really get deployed. There is no driver log at all in this test case which indicate all the containers are not started.
```Jul 31 11:12:56.182: INFO: Waiting up to 5m0s for PersistentVolumeClaims [pvc-lgpcg] to have phase Bound```
The system was expecting to get the PVC on 11:12. However, a look at `kubelet` log I found something like:
```
Jul 31 11:18:26 kind-worker kubelet[694]: I0731 11:18:26.505917     694 reconciler.go:319] Volume detached for volume "socket-dir" (UniqueName: "kubernetes.io/host-path/fda5159d-5404-49c7-be7b-0217233d9ea4-socket-dir") on node "kind-worker" DevicePath ""
...........
Jul 31 11:18:30 kind-worker kubelet[694]: E0731 11:18:30.267971     694 pod_workers.go:191] Error syncing pod fda5159d-5404-49c7-be7b-0217233d9ea4 ("csi-mockplugin-0_csi-mock-volumes-7860(fda5159d-5404-49c7-be7b-0217233d9ea4)"), skipping: failed to "StartContainer" for "mock" with CreateContainerConfigError: "cannot find volume \"socket-dir\" to mount into container \"mock\""
```
The kubelet only try to start the container on 11:18 which is already timeout for the test case. And it is no wonder it gets a "volume not found error" because after the test failed we have deleted the statefulset and thus the volume was unattached.

I am not sure how does this work, but apparently a lot of these timeout leads to the same root cause which is the system assume the mock csi driver pod has already started and start to wait for the PVC creation. However, what actually happen is the pod seems to be scheduled but the container for those pods for some reason are pretty slow to start. Two different examples:
1. https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-kind/1286019551993008128
In this case either the mock-csi-driver container is too slow to start or the mock csi driver register the socket on the node which is too slow. I am suspecting the latter because the other containers are able to bring up successfully. So there is no reason this container should fail alone. One follow-up we can do is to add timestamp on the mock CSI driver to understand the timeline better.
2. https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-kind/1289150375802179587
In this case the pod being scheduled but all the containers are not able to start before the timeout. As per the log information I found above.

I am start to thinking if it is possible that we run too many tests in parallel and the node is not able to handle too much workload and become slow to response? I would appreciate if anyone can give me some pointer on that front.

**Comment by Jiawei0227** [comment]
A look at the kind cluster, I found most of the time csi-provisioner is able to talk to the socket pretty fast. Only 2 of around 20 test cases I looked took around 1-3 minutes to connect. And 1 of them wait for around 5 minutes to connect the socket which also leads to test failure.

**Comment by Jiawei0227** [comment]
One thing I find also need to be improved is:
We are not waiting for the CSI driver pod actually being deployed to do the work. Instead, we are only checking if the CSIDriver object being created:
https://github.com/kubernetes/kubernetes/blob/0051d65f9f30db724dfb88256f70c23e37f6b257/test/e2e/storage/csi_mock_volume.go#L1427

This seems faulty because the CSIDriver object does not mean the driver pod is actually ready to perform the test.

**Comment by msau42** [comment]
The fix that @Jiawei0227 made in #93618 will wait for the csi driver to be installed on the node, before starting the rest of the test. This should mitigate some of the timing failures, but does mask the issue that it sometimes takes > 5 minutes to start these pods. We may see the tests increase in duration instead of failing.
